chore: add logic in underlying libraries for job logs

bvtujo opened the pull request:
- chore: add job logs logic and validation
- chore: update cloudwatchlogs to allow limiting log streams fetched and custom prefixes
- chore: rename service.go to workload.go in logging pkg
- chore: add logic for fetching state machine logs to workload
- chore: add unit tests for workload logging pkg

Adds unit tests and changes for the cloudwatchlogs and logging libraries necessary for job logs, specifically for getting one or N executions at a time, optionally including state machine logs. 

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

bvtujo committed:
chore: add job logs logic and validation

bvtujo committed:
chore: update cloudwatchlogs to allow limiting log streams fetched and custom prefixes

bvtujo committed:
chore: rename service.go to workload.go in logging pkg

bvtujo committed:
chore: add logic for fetching state machine logs to workload

bvtujo committed:
chore: add unit tests for workload logging pkg

bvtujo committed:
chore: make things work for ci

huanjani commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs_test.go:
```diff
@@ -237,6 +238,146 @@ func TestLogEvents(t *testing.T) {
+					OrderBy:      aws.String("LastEventTime"),
+					Limit:        aws.Int64(1),
+				}).Return(&cloudwatchlogs.DescribeLogStreamsOutput{
+					LogStreams: []*cloudwatchlogs.LogStream{
```
I may be reading this wrong, but as in the next test, should this return more than one log stream so that we can confirm that only one was fetched?

huanjani commented on internal/pkg/cli/job_logs.go:
```diff
@@ -23,6 +23,9 @@ import (
 	jobAppNamePrompt = "Which application does your job belong to?"
+
+	jobLogNamePrompt     = "Which job's logs would you like to show?"
+	jobLogNameHelpPrompt = "The logs of a deployed job will be shown."
```
nit: maybe something like this, bc this is the name prompt?
```suggestion
	jobLogNameHelpPrompt = "The logs of the indicated deployed job will be shown."
```

iamhopaul123 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -46,11 +46,13 @@ type LogEventsOutput struct {
 type LogEventsOpts struct {
 	LogGroup            string
-	LogStreams          []string // If nil, retrieve logs from all log streams.
+	LogStreamPrefixes   []string // If nil, retrieve logs from all log streams.
```
Maybe
```suggestion
	LogStreamPrefixesFilter   []string
```

iamhopaul123 commented on internal/pkg/logging/workload.go:
```diff
@@ -49,18 +51,22 @@ type WriteLogEventsOpts struct {
 	TaskIDs   []string
 	// OnEvents is a handler that's invoked when logs are retrieved from the service.
 	OnEvents func(w io.Writer, logs []HumanJSONStringer) error
+	// LogStreamLimit is an optional parameter for jobs and tasks to speed up CW queries
+	// involving multiple log streams.
+	LogStreamLimit *int64
```
Is this going to be a new flag or how do users specify this?

bvtujo replied:
this is specified by the `--last` flag; as in `job logs for the last N executions`.

iamhopaul123 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -60,13 +62,15 @@ func New(s *session.Session) *CloudWatchLogs {
 		Descending:   aws.Bool(true),
 		OrderBy:      aws.String(cloudwatchlogs.OrderByLastEventTime),
-	})
+		Limit:        logStreamLimit,
```
I feel like we should do pagination and do the filtering for each page and then apply this limit after that (quit querying the next page if limit is reached). Otherwise it could happen that `logStreamLimit` is 20 whereas the none of the first 20 log streams with the latest events matches the log stream prefixes, resulting no log stream returned.

bvtujo committed:
chore: address feedback

use pagination for log stream fetching.
rename variables.
make sure log stream truncation and filtering are engaged in tests.

iamhopaul123 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -60,38 +62,55 @@ func New(s *session.Session) *CloudWatchLogs {
+		} else {
+			logStreamNames = append(logStreamNames, streams...)
+		}
+		if len(logStreamNames) > logStreamLimit {
```
nit
```suggestion
		if len(logStreamNames) >= logStreamLimit {
```

iamhopaul123 commented on internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go:
```diff
@@ -147,6 +166,13 @@ func truncateEvents(limit int, events []*Event) []*Event {
+	if len(streams) <= limit {
+		return streams
+	}
+	return streams[:len(streams)-limit]
```
Should this be
```suggestion
	return streams[:limit]
```

bvtujo committed:
Update internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go

Co-authored-by: Penghao He <penghaoh@amazon.com>

bvtujo committed:
chore: fix logic when using int, not int64

huanjani reviewed (approved):
LGTM!

iamhopaul123 reviewed (approved):
🚢

mergify[bot] committed:
Merge branch 'mainline' into job-logger
